MdePkg/Library/BaseLib/ConvertGuidUuid.c

```diff
@@ -0,0 +1,131 @@
+
+  UUIDs contains 5 fields:
+  - time_low: 32 bits
+  - time_mid: 16 bits
+  - time_hi_and_version: 16 bits
+  - clock_seq_hi_and_reserved: 8 bits
+  - clock_seq_low: 8 bits
+  - node: 48 bits
+
+  Each field encoded with the Most Significant Byte first (known as network byte
+  order).
+
+  GUID (Globally Unique Identifier), on the other hand, is a 128-bit number used
+  in UEFI environments, which is similar to UUID but has a different byte order
+  in memory. See https://uefi.org/specs/UEFI/2.11/Apx_A_GUID_and_Time_Formats.html
+
+  GUID also contains 5 fields:
+  - TimeLow: 32 bits
+  - TimeMid: 16 bits
+  - TimeHiAndVersion: 16 bits
+  - ClockSeqHighAndReserved: 16 bits
+  - ClockSeqLow: 8 bits
+  - Node: 48 bits
+
+  TimeLow, TimeMid, TimeHighAndVersion fields in the EFI are encoded as little endian.
+
+  Example:
+  Consider the same string representation/registry format for MM communication v2:
+  "378daedc-f06b-4446-8314-40ab933c87a3"
+
+  In UUID format, it is represented as:
+  - Data fields:
+    - time_low: 0x37 0x8d 0xae 0xdc (0x378daedc in big-endian)
+    - time_mid: 0xf0 0x6b (0xf06b in big-endian)
+    - time_hi_and_version: 0x44 0x46 (0x4446 in big-endian)
+    - clock_seq_hi_and_reserved: 0x83
+    - clock_seq_low: 0x14
+    - node: 0x00, 0xab, 0x93, 0x3c, 0x87, 0xa3
+  - Byte representation in memory:
+    - 37 8d ae dc f0 6b 44 46 83 14 40 ab 93 3c 87 a3
+
+  However, in GUID format, it is represented as:
+  - Data fields:
+    - TimeLow: 0xdc 0xae 0x8d 0x37 (0x378daedc in little-endian)
+    - TimeMid: 0x6b 0xf0 (0xf06b in little-endian)
+    - TimeHiAndVersion: 0x46 0x44 (0x4446 in little-endian)
+    - ClockSeqHighAndReserved: 0x83
+    - ClockSeqLow: 0x14
+    - Node: 0x00, 0xab, 0x93, 0x3c, 0x87, 0xa3
+  - Byte representation in memory:
+    - dc ae 8d 37 6b f0 46 44 83 14 40 ab 93 3c 87 a3
+
+*******************************************************************************/
+
+/**
+  This function converts a GUID to a UUID in RFC4122 format.
+
+  The conversion is done by swapping the byte order of the TimeLow, TimeMid, and
+  TimeHiAndVersion fields, while keeping the ClockSeq and Node fields unchanged.
+
+  @param [in] FromGuid  GUID to be converted to UUID format.
+                        The GUID is expected to be in the UEFI spec defined format,
+                        which is little-endian for TimeLow, TimeMid, and TimeHiAndVersion
+                        fields.
+  @param [out] ToUuid   Pointer to a UINT64 array that will hold the converted UUID.
+                        The UUID will be in the RFC4122 format, which is big-endian for
+                        time_low, time_mid, and time_hi_and_version fields.
+**/
+VOID
+EFIAPI
+ConvertGuidToUuid (
+  IN   GUID    *FromGuid,
+  OUT  UINT64  *ToUuid
+  )
+{
+  UINT32  *Data32;
+  UINT16  *Data16;
+
+  CopyGuid ((GUID *)ToUuid, FromGuid);
+  Data32    = (UINT32 *)ToUuid;
+  Data32[0] = SwapBytes32 (Data32[0]);

```

Comment:
Why not call SwapBytesxx() on the fields of GUID without using additional local variables?

Also, why is UUID a UINT64*?  Can't is also be type GUID?  The field names are the same for UUID and GUID.  Only the endianness is different.

Reply:
I do not agree.  UINT64* is not a UUID by type or size.  Look at the python uuid module as an example.  It uses the same object class for both forms and has a single method to return the little endian version of an uuid object.

Reply:
LGTM.  Thanks for the quick updates.